src/iidm/util/TerminalFinder.cpp

```diff
@@ -0,0 +1,65 @@
+/**
+ * Copyright (c) 2020, RTE (http://www.rte-france.com)
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include <powsybl/iidm/util/TerminalFinder.hpp>
+
+#include <powsybl/iidm/Branch.hpp>
+#include <powsybl/iidm/BusbarSection.hpp>
+#include <powsybl/iidm/HvdcConverterStation.hpp>
+#include <powsybl/iidm/Injection.hpp>
+#include <powsybl/iidm/TwoWindingsTransformer.hpp>
+#include <powsybl/stdcxx/instanceof.hpp>
+
+namespace powsybl {
+
+namespace iidm {
+
+namespace util {
+
+TerminalFinder::TerminalFinder(const std::vector<Predicate>& predicates) :
+    m_comparator(getComparator(predicates)) {
+}
+
+stdcxx::Reference<Terminal> TerminalFinder::find(const stdcxx::range<Terminal>& terminals) const {
+    auto it = std::max_element(terminals.begin(), terminals.end(), m_comparator);
+    return it != terminals.end() ? stdcxx::ref(*it) : stdcxx::ref<Terminal>();
+}
+
+TerminalFinder::Comparator TerminalFinder::getComparator(const std::vector<Predicate>& predicates) {
+    return [&predicates](Terminal& t1, Terminal& t2) {
+        if (stdcxx::areSame(t1, t2)) {
+            return false;
+        }
+        for (const auto& predicate : predicates) {
+            if (predicate(t1) || predicate(t2)) {
+                return !predicate(t1) && predicate(t2);
+            }
```

Comment:
This code is more complex to understand than the java version. The comments have been removed and I'm not this is equivalent. Is there a good reason to write it like this?

Reply:
Agree with you! Actually, the problem is that a C++ comparator returns a boolean, unlike in Java where it may return an (negative, 0 or positive) integer. 

In C++, returning -1 is equivalent to true, that's why Java implementation could not be copied/paste here.

I copied/pasted Java test suite (see `TerminalFinderTest.cpp`), so that we check that we stricltly have the same behaviour.